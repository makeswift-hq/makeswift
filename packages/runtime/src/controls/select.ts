export const SelectControlType = 'makeswift::controls::select'

type SelectControlOption<T extends string> = { value: T; label: string }

type SelectControlConfig<T extends string = string> = {
  label?: string
  labelOrientation?: 'horizontal' | 'vertical'
  options: SelectControlOption<T>[]
  defaultValue?: T
}

export type SelectControlDefinition<C extends SelectControlConfig = SelectControlConfig> = {
  type: typeof SelectControlType
  config: C
}

export type SelectControlDefinitionOption<T extends SelectControlDefinition> =
  T['config'] extends SelectControlConfig<infer U> ? U : never

export type SelectControlData<T extends SelectControlDefinition = SelectControlDefinition> =
  SelectControlDefinitionOption<T> | null

export function Select<T extends string, C extends SelectControlConfig<T>>(
  config: C & { options: SelectControlOption<T>[] },
): SelectControlDefinition<C> {
  return { type: SelectControlType, config }
}
