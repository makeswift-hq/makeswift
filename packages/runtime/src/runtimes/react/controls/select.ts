import { type DataType, type ResolvedValueType } from '@makeswift/controls'

import { SelectDefinition } from '../../../controls'

export function useSelectControlValue(
  data: DataType<SelectDefinition> | undefined,
  definition: SelectDefinition,
): ResolvedValueType<SelectDefinition> {
  return definition.fromData(data) ?? definition.config.defaultValue
}
