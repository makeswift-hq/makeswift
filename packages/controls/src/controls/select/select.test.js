import { SelectDefinition } from './select'

describe('Select', () => {
  describe('deserialize', () => {
    test('string values', () => {
      expect(
        SelectDefinition.deserialize({
          type: SelectDefinition.type,
          config: {
            label: 'Block type',
            labelOrientation: 'horizontal',
            options: [
              { value: 'p', label: 'Paragraph' },
              { value: 'h1', label: 'Heading 1' },
            ],
            defaultValue: 'p',
          },
        }),
      ).toMatchSnapshot()
    })

    test('correctly coerces numeric values to string', () => {
      expect(
        SelectDefinition.deserialize({
          type: SelectDefinition.type,
          config: {
            label: 'Text size',
            options: [
              { value: 8, label: 'sm' },
              { value: 12, label: 'md' },
              { value: 16, label: 'lg' },
            ],
            defaultValue: 12,
          },
        }),
      ).toMatchSnapshot()
    })
  })
})
