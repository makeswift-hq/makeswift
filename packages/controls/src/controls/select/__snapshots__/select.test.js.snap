// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select deserialize correctly coerces numeric values to string 1`] = `
SelectDefinition {
  "config": {
    "defaultValue": "12",
    "label": "Text size",
    "options": [
      {
        "label": "sm",
        "value": "8",
      },
      {
        "label": "md",
        "value": "12",
      },
      {
        "label": "lg",
        "value": "16",
      },
    ],
  },
}
`;

exports[`Select deserialize string values 1`] = `
SelectDefinition {
  "config": {
    "defaultValue": "p",
    "label": "Block type",
    "labelOrientation": "horizontal",
    "options": [
      {
        "label": "Paragraph",
        "value": "p",
      },
      {
        "label": "Heading 1",
        "value": "h1",
      },
    ],
  },
}
`;
